src/persist-client/src/read.rs

```diff
@@ -123,21 +140,36 @@ where
     /// [differential_dataflow::consolidation::consolidate_updates].
     ///
     /// An None value is returned if this iterator is exhausted.
-    #[instrument(level = "debug", name = "snap::next", skip_all, fields(shard = %self.shard_id))]
+    #[instrument(level = "debug", name = "snap::next", skip_all, fields(shard = %self.handle.machine.shard_id()))]
     pub async fn next(&mut self) -> Option<Vec<((Result<K, String>, Result<V, String>), T, D)>> {
         trace!("SnapshotIter::next");
+
         loop {
             let (key, desc) = match self.batches.pop() {
-                Some(x) => x,
-                // All done!
-                None => return None,
+                Some(x) => {
+                    // Periodically heartbeat so the batches we're iterating
+                    // don't get garbage collected.
+                    self.handle.maybe_heartbeat(None).await;
+                    x
+                }
+                // All done! Drop our seqno capability to unblock garbage
+                // collection of blobs.
+                None => {
+                    trace!("SnapshotIter::expire");
+                    self.handle
+                        .machine
+                        .expire_reader(&self.handle.reader_id)
+                        .await;
+                    self.handle.explicitly_expired = true;
```

Comment:
will expiring the handle here cause problems for `persist_source`, where the handle is reused for the Listen?

Reply:
oh! nvm, this is a separate handle, one created by snapshot splits